### PR TITLE
Add sq, cq, resources for queue and regex_dev, enqueue and dequeue functionalities and test

### PR DESCRIPTION
**Pull request by vuhuong**
This PR includes:
+ Add CQ create, destroy and arm helper APIs
+ Add SQ create and destroy helper APIs
+ Implement QP create callback
+ Add device and queue resources for enqueue dequeue
+ Add and enable enqueue and dequeue functions
+ Add test regex appliacation

### CODE REVIEW

File: drivers/regex/mlx5/mlx5_regex_queue.c [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+	DEVX_SET(regexp_mmo_control, seg, subset_id_2, ops->group_id2);
+	DEVX_SET(regexp_mmo_control, seg, subset_id_3, ops->group_id3);
+}
+
+static __rte_always_inline void
+mlx5dv_set_metadata_seg(struct mlx5_wqe_metadata_seg *seg,
+			uint32_t mmo_control_31_0, uint32_t lkey,
+			uintptr_t address)
+{
+	seg->mmo_control_31_0 = htobe32(mmo_control_31_0);
```
Comment (orikam) [inline drivers/regex/mlx5/mlx5_regex_queue.c]: should be changed to rte_cpu_to_be_32
same for all 3 lines

File: drivers/regex/mlx5/mlx5_regex_queue.c [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+	mlx5dv_set_ctrl_seg(cseg, q->sq.pi, MLX5_OPCODE_MMO,
+			    MLX5_OPC_MOD_MMO_REGEX, q->sq.sqn,
+			    MLX5_WQE_CTRL_CQ_UPDATE, nsegs, 0,
+			    regex_cseg.le_subset_id_0_subset_id_1);
+	/* Set wqe metadata segment */
+	mseg = (struct mlx5_wqe_metadata_seg *)((uint8_t*)cseg + sizeof(*cseg));
+	lkey = mlx5_regex_mr_addr2mr_bh(dev, &q->mr_ctrl,
+					job->m_iov.buf_iova);
+	mlx5dv_set_data_seg(&tmp_dseg, job->m_iov.buf_size,
+			    lkey, job->m_iov.buf_iova);
```
Comment (orikam) [inline drivers/regex/mlx5/mlx5_regex_queue.c]: we should create those function in DPDK.
buf_iova should be changed to buf addr

**Review by orikam — CHANGES_REQUESTED** [review]
I will review more later

File: drivers/regex/mlx5/mlx5_regex_queue.c [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+			    lkey, (uintptr_t)BUF_ADDR((*ops->bufs)[0]));
+	/* Data segment for response buffer */
+	dseg += sizeof(*dseg);
+	lkey = mlx5_regex_mr_addr2mr_bh(dev, &q->mr_ctrl,
+					(uintptr_t)BUF_ADDR(&job->resp_iov));
+	mlx5dv_set_data_seg(dseg, BUF_SIZE(&job->resp_iov),
+			    lkey, (uintptr_t)BUF_ADDR(&job->resp_iov));
+	/* Ring doorbell */
+	rte_io_wmb();
+	q->sq.db_rec[MLX5_SND_DBR] = rte_cpu_to_be_32(q->sq.pi & MLX5_QP_CI_MASK);
```
Comment (yuvalavn) [inline drivers/regex/mlx5/mlx5_regex_queue.c:120]: Hi vu
this a bug
please see PRM
need to be the next empty wqbb and not wqe index

File: drivers/regex/mlx5/mlx5_regex_queue.c [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+	rte_wmb();
+#ifdef RTE_ARCH_64
+	*(uint64_t *)q->sq.db_addr = *(uint64_t *)cseg;
+#else
+	*(uint32_t *)q->sq.db_addr = *(uint64_t *)cseg;
```
Comment (yuvalavn) [inline drivers/regex/mlx5/mlx5_regex_queue.c]: Every write need to be alternated between odd and even on the uar
please see PARM

File: drivers/regex/mlx5/mlx5_regex_queue.c [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+	lkey = mlx5_regex_mr_addr2mr_bh(dev, &q->mr_ctrl,
+					(uintptr_t)BUF_ADDR(&job->resp_iov));
+	mlx5dv_set_data_seg(dseg, BUF_SIZE(&job->resp_iov),
+			    lkey, (uintptr_t)BUF_ADDR(&job->resp_iov));
+	/* Ring doorbell */
+	rte_io_wmb();
+	q->sq.db_rec[MLX5_SND_DBR] = rte_cpu_to_be_32(q->sq.pi & MLX5_QP_CI_MASK);
+	rte_wmb();
+#ifdef RTE_ARCH_64
+	*(uint64_t *)q->sq.db_addr = *(uint64_t *)cseg;
```
Comment (yuvalavn) [inline drivers/regex/mlx5/mlx5_regex_queue.c]: We should use blueflame for performance
It means writing the whole wqe to the BAR not just the ctrl segment

File: drivers/regex/mlx5/mlx5_regex_queue.c (hunk 4 above)
Comment (yuvalavn) [reply drivers/regex/mlx5/mlx5_regex_queue.c]: this comment relates to blueflame

File: drivers/regex/mlx5/mlx5_regex_queue.c (hunk 5 above)
Comment (yuvalavn) [inline drivers/regex/mlx5/mlx5_regex_queue.c]: Should be
*(volatile uint64_t *)cseg;

**Review by yuvalavn — CHANGES_REQUESTED** [review]

**Pull request closed without merging**